### ISSUE TITLE
Can't delete single email from GUI

### ISSUE THREAD

**Issue by alexrussell** [sj26/mailcatcher#232]
When I try to use the 'delete' key (actually backspace as I'm on a Mac) to delete a single message in the web GUI it always fails. I can use 'clear' to delete all, but single ones never delete.

I am using 0.6.1 according to `gem list`.

FWIW it does a DELETE request to `/messages/1` and doesn't seem to pass anything else (query string or request body) and the server fails with a 500 and, apparently the following full response: `<h1>Internal Server Error</h1>`

Full details, from Chrome's network inspector:

### General
- Remote Address: `192.168.0.26:1080`
- Request URL: `http://mailcatcher:1080/messages/1`
- Request Method: `DELETE`
- Status Code: `500 Internal Server Error`

### Request Headers

    DELETE /messages/1 HTTP/1.1
    Host: mailcatcher:1080
    Connection: keep-alive
    Pragma: no-cache
    Cache-Control: no-cache
    Accept: */*
    Origin: http://mailcatcher:1080
    X-Requested-With: XMLHttpRequest
    User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/43.0.2357.124 Safari/537.36
    DNT: 1
    Referer: http://mailcatcher:1080/
    Accept-Encoding: gzip, deflate, sdch
    Accept-Language: en,en-US;q=0.8

### Response Headers

    HTTP/1.1 500 Internal Server Error
    Content-Type: text/html;charset=utf-8
    Content-Length: 30
    X-XSS-Protection: 1; mode=block
    X-Content-Type-Options: nosniff
    X-Frame-Options: SAMEORIGIN
    Connection: keep-alive
    Server: thin 1.5.1 codename Straight Razor

Looking further into it (and your curl-based tests on #102), I get the same results outside of Chrome:

```
→ curl http://mailcatcher:1080/messages
[{"id":1,"sender":"\u003credacted@redacted.com\u003e","recipients":["\u003credacted@redacted.com\u003e"],"subject":"Subject","size":"4891","created_at":"2015-06-22T14:25:21.000+00:00"}]
→ curl -XDELETE http://mailcatcher:1080/messages/1
<h1>Internal Server Error</h1>
```

And yes, a request to `/messages/1.json` returns the full message in JSON form. `/messages/1.html` returns the HTML of the message.

Related but not the same (as they return 404-like responses): #102, #149, #169

**Comment by MacroMackie** [comment]
I also seem to be experiencing similar issues while using DELETE requests:

```
curl -XGET http://localhost:1080/messages/1.source
MIME-Version: 1.0
Date: 2015-07-29 11:29:06 -0400
Message-ID: <63154712.136.1438183747128.JavaMail.root@be80daa0dd10>
Subject: Default subject line.
```

```
curl -XDELETE http://localhost:1080/messages/1
SQLite3::RangeException: bind or column index out of range
	/var/lib/gems/1.9.1/gems/sqlite3-1.3.10/lib/sqlite3/statement.rb:41:in `bind_param'
	/var/lib/gems/1.9.1/gems/sqlite3-1.3.10/lib/sqlite3/statement.rb:41:in `block in bind_params'
	/var/lib/gems/1.9.1/gems/sqlite3-1.3.10/lib/sqlite3/statement.rb:37:in `each'
	/var/lib/gems/1.9.1/gems/sqlite3-1.3.10/lib/sqlite3/statement.rb:37:in `bind_params'
	/var/lib/gems/1.9.1/gems/sqlite3-1.3.10/lib/sqlite3/statement.rb:64:in `execute'
	/var/lib/gems/1.9.1/gems/mailcatcher-0.5.12/lib/mail_catcher/mail.rb:157:in `delete_message!'
	/var/lib/gems/1.9.1/gems/mailcatcher-0.5.12/lib/mail_catcher/web.rb:138:in `block in <class:Web>'
	...                                             
```

I am currently running mailcatcher in a docker container, which may be related to this issue? It seems strange the get requests are fine for specific messages, but the delete requests are not.

This is the method that is failing:
```ruby
delete "/messages/:id" do
  id = params[:id].to_i
  if message = Mail.message(id)
    Mail.delete_message!(id)
    status 204
  else
    not_found
  end
end
```

**Comment by MacroMackie** [comment]
Looks like this was related to the issue:
```ruby
def delete!
    @delete_messages_query ||= db.prepare "DELETE FROM message"
    @delete_message_parts_query ||= db.prepare "DELETE FROM message_part"

    @delete_messages_query.execute and
    @delete_message_parts_query.execute
  end
```
```ruby
  def delete_message!(message_id)
    @delete_messages_query ||= db.prepare "DELETE FROM message WHERE id = ?"
    @delete_message_parts_query ||= db.prepare "DELETE FROM message_part WHERE message_id = ?"
    @delete_messages_query.execute(message_id) and
    @delete_message_parts_query.execute(message_id)
```
`@delete_messages_query` and `@delete_message_parts_query` are shared across both delete! and delete_message!, which may have been what was causing the `SQLite3::RangeException: bind or column index out of range`

Switching the former to `@delete_all_messages_query` and `@delete_all_message_parts_query` seems to have fixed the issue for me.

I'll continue testing and see if the old issue still occurs.

**Comment by alexrussell** [comment]
I don't know if I have updated mailcatcher and it's now working, but I can now delete singly or a batch 'purge' from the UI. As such, I'm closing this issue.

**Comment by alexrussell** [comment]
(Just checked - I haven't updated, still 0.6.1. Weird!)

**Comment by timothyasp** [comment]
I'm still having this problem, where making a request to delete a single message fails with a 500.

**Comment by alexrussell** [comment]
You know what, I now have this issue again. As such, I'll reopen but obviously we're very much needing a reproducible test case here. How can I have the problem, it goes away, and then it comes back?

**Comment by timothyasp** [comment]
I've forked it and am working on a fix.

**Comment by duduribeiro** [comment]
@alexrussell still with the problem? I tested on master and on v0.6.1 (0179a92) and worked fine for me =/

Thanks

**Comment by alexrussell** [comment]
I get it on and off. Sometimes both deletion actions work, sometimes delete single is the only command that works (whereas clear all doesn't) and sometimes the other way around. But it feels like this is relatively impossible to work out, so I dunno how helpful I can be.

**Comment by sj26** [comment]
This should be fixed by #258 thanks to @grant-mccarriagher, included in v0.6.3. Please update and comment again if you have issues.
